components/NavPage.qml

```diff
@@ -0,0 +1,201 @@
+                noResult.visible = true;
+            } else {
+                noResult.visible = false;
+            }
+        }
+    }
+
+    SplashComponent {
+        id: httpFailedSearch
+        objectName: "HTTPFailedSearch"
+    }
+
+    TextField {
+        id: startPoint
+        width: parent ? parent.width - units.gu(2) : undefined
+        hasClearButton: true
+        focus: true
+        placeholderText: i18n.tr("Enter a place name to start from")
+        primaryItem: Image {
+            height: parent.height*0.5
+            width: parent.height*0.5
+            anchors.verticalCenter: parent.verticalCenter
+            anchors.verticalCenterOffset: -units.gu(0.2)
+            source: Helpers.getIcon("search")
+        }
+        onAccepted: {
+            if (text) {
+                navState = "startpoint"
+                searchModel.searchString = text;
+                searchModel.search();
+                startPoint.focus = false
+            }
+        }
+        // Indicator to show search activity
+        ActivityIndicator {
+            id: startSearchActivity
+            anchors {
+                right: startPoint.right
+                rightMargin: units.gu(1)
+                verticalCenter: startPoint.verticalCenter
+            }
+            running: (navState === "startpoint" && searchModel.status === XmlListModel.Loading)
+        }
+    }
+    TextField {
+        id: endPoint
+        anchors.top: startPoint.bottom
+        width: parent ? parent.width - units.gu(2) : undefined
+        hasClearButton: true
+        placeholderText: i18n.tr("Enter a place name to go to")
+        primaryItem: Image {
+            height: parent.height*0.5
+            width: parent.height*0.5
+            anchors.verticalCenter: parent.verticalCenter
+            anchors.verticalCenterOffset: -units.gu(0.2)
+            source: Helpers.getIcon("search")
+        }
+        onAccepted: {
+            if (text) {
+                navState = "endpoint"
+                searchModel.searchString = text;
+                searchModel.search();
+                endPoint.focus = false
+            }
+        }
+        // Indicator to show search activity
+        ActivityIndicator {
+            id: endSearchActivity
+            anchors {
+                right: endPoint.right
+                rightMargin: units.gu(1)
+                verticalCenter: endPoint.verticalCenter
+            }
+            running: (navState === "endpoint" && searchModel.status === XmlListModel.Loading)
+        }
+    }
+    Button {
+        id: carNavBtn
+        text: i18n.tr("By Car")
+        anchors.top: endPoint.bottom
+        onClicked: {
+            map.navigate(fromlat, fromlon, tolat, tolon, RouteQuery.CarTravel);
+            map.goToLatLng(fromlat, fromlon);
+            navPage.pageStack.pop();
+        }
+    }
+// XXX: OSRM doesn't support these routing methods. :(
```

Comment:
Funny story: before I realised QtLocation provided access to OSRM routing, I started out trying to write a QML component to provide this interface. I got lost, because I'm still finding my way around with QML, but I also thought of MapQuest and Grasshopper. And then when I came to get the turn-by-turn directions from OSRM - the OSRM server is returning route_information that has nine elements per segment, where QtLocation is expecting eight, so the directions never appear! So writing that QML component might be needed after all!